### ISSUE TITLE
Improving the rendering performance

### ISSUE THREAD

**Issue by vy** [vy/log4j2-logstash-layout#17]
We think we're going to be able to open-source our message class fairly soon, at which point it might be interesting to explore alternate layout/message interfaces that reduce serialization overhead.  Maybe something along the lines of a visitor pattern, where the Layout passes a visitor to the Message and the Message does the tree-walk?  Some people refer to that as a SAX-style API, and Jackson's "Streaming API" (JsonGenerator, https://fasterxml.github.io/jackson-core/javadoc/2.9/com/fasterxml/jackson/core/JsonGenerator.html) works that way.

Have you worked at all with log4j2 upstream?  It seems like this system would benefit from some hard-core perfomance work, which generally involves some API refactoring.  If we could get a team together, it would be satisfying to build a "fish tagging" system for Java distributed architectures, comparable to what I've helped build previously in a Javascript/Python/C++ world.  (That system resembled log4j2's ContextDataInjector scheme, with the local context transferred as metadata on RESTful API requests and auto-captured by task closures for delegation across thread boundaries.)  Netty-based services really need such a thing.

_Originally posted by @mkedwards in https://github.com/vy/log4j2-logstash-layout/pull/15#issuecomment-428045091_

**Comment by mkedwards** [comment]
For discussion's sake, this is the bit of code that could be reworked to use a `JsonGenerator` instead of an `ObjectWriter`:

```
 86     private String render(LogEvent event, ObjectNode srcRootNode) {
 87 
 88         // Resolve node and bail-out if empty.
 89         JsonNode dstRootNode = resolveNode(event, srcRootNode);
 90         if (dstRootNode.isNull()) {
 91             return "{}";
 92         }
 93 
 94         // Serialize resolved node.
 95         try {
 96             ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
 97             try (OutputStreamWriter writer = new OutputStreamWriter(outputStream, JSON_CHARSET)) {
 98                 objectWriter.writeValue(writer, dstRootNode);
 99                 writer.write(System.lineSeparator());
100             }
101             return outputStream.toString(JSON_CHARSET_NAME);
102         } catch (IOException error) {
103             throw new RuntimeException("failed serializing JSON", error);
104         }
105 
106     }
```

It would be ideal not to have to create the `dstRootNode` intermediate object at all.  As a halfway step to that, I think it would be useful to subclass the `MultiformatMessage` interface (or add an interface mixin or something) to add a `streamMessageTo(JsonGenerator out)` method.  Pass it an instance `buffer` of `TokenBuffer`, and then use `objectMapper.readTree(buffer.asParser())` to get a `JsonNode` that can be inserted into `dstRootNode`.

**Comment by vy** [comment]
TODOs:
- Accept `ObjectMapper` as a parameter in `LogstashLayout.Builder`.
- Use Jackson Afterburner module.

**Comment by vy** [comment]
Adding Afterburner is not improving anything -- makes sense, we are not doing any POJO serialization. Long live benchmarks!

**Comment by mkedwards** [comment]
I love what you're doing, and there's no need to apologize for the pace.  We're going through the open-sourcing motions here, should have something out that uses the Streamable API shortly.

**Comment by mkedwards** [comment]
Public now:  https://github.com/nexiahome/formatted_data_message/tree/logstash-streamable

**Comment by mkedwards** [comment]
The other thing that we found to be very important in the previous project was readability/familiarity of the syntactic sugar at the log site.  In C++ that was handled with macro/template magic to interpolate tagged values into a template string, with automagical fallbacks for untagged parameters, all wrapped up in a syntax identical to the Google-logging library we were already using.  We'll want to do something similar in Java-space, probably involving horrible generic/reflection machinery.  I'm not that clever at Java 8+ yet, but I fear that's going to have to change.

**Comment by mkedwards** [comment]
Very cool!  Thank you, and please let me know how I can help.  I will be baking this into our Netty-based system soon -- we're currently sorting out thread affinity stuff -- and should be able to produce a relevant parallel benchmark that spews per-message log traffic into kafka for later aggregation into per-device traffic stats.

**Comment by vy** [comment]
Done. In `json-generator` branch, see the benchmark results reported in `README.md`. I would really appreciate an in-detail review of the commit. If you approve, I will merge it to master and put the ribbon for a new release. For the records, I have GiST'ed [the output of `LogstashLayoutBenchmark`](https://gist.github.com/vy/0ebb6a4ad020f115b30edd42ba086525) on my machine.

**Comment by mkedwards** [comment]
I haven't forgotten about this.  Getting to it now.  Just had to rework a library under our Netty thing first, so I could get a high-volume logger where I wanted it.

**Comment by vy** [comment]
Just finished wrapping up the GC-free version of the layout. I increased the performance up to 1,109,501 msg/sec with 0.645 KB/sec alloc rate. I am busy with spitting out a plot after the JMH benchmark runs.

**Comment by vy** [comment]
@mkedwards, @emschwar, I am done with my changes. Would you mind taking a look when you have time, please?

**Comment by vy** [comment]
@mkedwards, hmmm..., no, we do not listen configuration changes. Let me check that. But please do review the rest at your convenience.

**Comment by mkedwards** [comment]
I'm having trouble combining the KafkaAppender and LogstashLayout.  That used to work, but it regressed recently.  Any idea offhand why?

**Comment by mkedwards** [comment]
To be precise, I'm using this branch of our fork of log4j2-logstash-layout:  https://github.com/nexiahome/log4j2-logstash-layout/tree/json-generator

and this branch of our streamable message class:  https://github.com/nexiahome/formatted_data_message/tree/logstash-streamable

and when I run `mvn test` in `formatted_data_message`, which logs some JSON-formatted messages to both console and kafka, the console version looks fine and the kafka version has no message body.  Something about the LogstashLayout being attached to the KafkaAppender doesn't work quite right.

I'll try to reproduce this without our forkage, but if you have any insight into where I might look, that would be quite helpful.

**Comment by mkedwards** [comment]
For context, here's what I'm currently getting out of the JSON formatter:
```
{
  "@version" : 1,
  "@timestamp" : "2018-11-07T11:49:27.774-08:00",
  "source_host" : "BitPushers-MacBook-Pro-4.local",
  "level" : "INFO",
  "logger_name" : "org.apache.logging.log4j.message.lazy.FormattedDataMessageTest",
  "thread_name" : "main",
  "message" : {
    "type" : "some_message_type",
    "id" : "some_message_id",
    "message" : "This is an exception message. a=aVal b=bVal",
    "template" : "This is an exception message. a=%(a) b=%(b)",
    "a" : "aVal",
    "b" : "bVal",
    "c" : "cVal"
  },
  "class" : "org.apache.logging.log4j.message.lazy.FormattedDataMessageTest",
  "method" : "throwerLogs",
  "file" : "FormattedDataMessageTest.java",
  "line_number" : 134,
  "exception" : {
    "exception_class" : "java.lang.RuntimeException",
    "exception_message" : "Nolan Ryan pitch",
    "stacktrace" : "java.lang.RuntimeException: Nolan Ryan pitch\n\tat org.apache.logging.log4j.message.lazy.FormattedDataMessageTest.throwerLogs(FormattedDataMessageTest.java:132)\n\tat java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)\n\tat java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)\n\tat java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)\n\tat java.base/java.lang.reflect.Method.invoke(Method.java:566)\n\tat org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:515)\n\tat org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:115)\n\tat org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$6(TestMethodTestDescriptor.java:171)\n\tat org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:72)\n\tat org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:167)\n\tat org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:114)\n\tat org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:59)\n\tat org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:105)\n\tat org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:72)\n\tat org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:95)\n\tat org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:71)\n\tat java.base/java.util.ArrayList.forEach(ArrayList.java:1540)\n\tat org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)\n\tat org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:110)\n\tat org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:72)\n\tat org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:95)\n\tat org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:71)\n\tat java.base/java.util.ArrayList.forEach(ArrayList.java:1540)\n\tat org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)\n\tat org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:110)\n\tat org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:72)\n\tat org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:95)\n\tat org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:71)\n\tat org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:32)\n\tat org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)\n\tat org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:51)\n\tat org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:220)\n\tat org.junit.platform.launcher.core.DefaultLauncher.lambda$execute$6(DefaultLauncher.java:188)\n\tat org.junit.platform.launcher.core.DefaultLauncher.withInterceptedStreams(DefaultLauncher.java:202)\n\tat org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:181)\n\tat org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:128)\n\tat org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invokeAllTests(JUnitPlatformProvider.java:142)\n\tat org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invoke(JUnitPlatformProvider.java:117)\n\tat org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:384)\n\tat org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:345)\n\tat org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:126)\n\tat org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:418)\n"
  }
}
{
  "@version" : 1,
  "@timestamp" : "2018-11-07T11:49:28.150-08:00",
  "source_host" : "BitPushers-MacBook-Pro-4.local",
  "level" : "INFO",
  "logger_name" : "org.apache.logging.log4j.message.lazy.FormattedDataMessageTest",
  "thread_name" : "main",
  "message" : {
    "type" : "a_message_type",
    "id" : "a_message_id",
    "message" : "This is a message. a=aVal b=bVal",
    "template" : "This is a message. a=%(a) b=%(b)",
    "a" : "aVal",
    "b" : "bVal",
    "c" : "cVal"
  },
  "class" : "org.apache.logging.log4j.message.lazy.FormattedDataMessageTest",
  "method" : "loggerLogs",
  "file" : "FormattedDataMessageTest.java",
  "line_number" : 125
}
```
Our goal is to be able to stream structured logging messages into kafka (or similar plumbing) and ingest them from there into a mostly column-oriented datastore.  Entries in the column-oriented datastore don't keep any of the non-timestamp high-cardinality columns, such as `message.message` or `exception.stacktrace`.  When you've got to see those columns, you go get the raw log message, looking up from a hash over a few columns that get you a reasonable degree of uniqueness (`@timestamp`, `source_host`, maybe `line_number`) into a document store.  But for most kinds of analysis, it's the `message.id` and the structured message parameters that you want to see, along with `@timestamp` and `source_host` and context parameters.  (That's where the `ContextDataInjector` comes in; we haven't written that part yet, but I hope to get to it soon, once we've got the ingestion part up and running.)

**Comment by mkedwards** [comment]
Okay, so what's going on here is that the Kafka appender wants `toByteArray()`.  You probably want to inherit from `AbstractStringLayout` instead of `Layout<String>` to get the fast conversion from the `String` returned by `toSerializable()` to a `byte[]`.  We've done this in our fork; I'll disentangle that from our version/artifactory noise and send you a separate PR.

**Comment by mkedwards** [comment]
It doesn't require it to extend from there; that just seemed like the smoothest way to get `toByteArray()`.  Tell me more about how gc-free rendering works, and let's figure out how to get `KafkaAppender` to use that path.

**Comment by vy** [comment]
**Disclaimer:** Following is my interpretation of the Log4j 2 source code. I am neither a Log4j 2 expert, nor committer, but just someone with quite some experience above the average.

Every *layout* is expected to implement `org.apache.logging.log4j.core.Layout<T extends Serializable>`, which extends `Encoder<LogEvent>`. There, the javadoc states the following: *Since 2.6, Layouts can encode (`Encoder#encode(Object, ByteBufferDestination)`) a `LogEvent` directly to a `ByteBufferDestination` without creating temporary intermediary objects.* That is, implementing `Layout<String>` in `LogstashLayout` is 1) sufficient for an interface compliant Log4j 2 *layout* and 2) provides access to garbage-free layout rendering via `Encoder#encode()`.

Assuming you are using the default `KafkaAppender` packaged with Log4j 2, I am really puzzled why it stopped working.

(Almost) all the default Log4j 2 layouts (`JSONLayout`, `XMLLayout`, etc.) extend from `AbstractStringLayout`, which implements `StringLayout`, which extends `Layout<String>` with an extra `Charset getCharset()` method. `AbstractStringLayout` contains the base set of utilities to ease `String` rendering, though these are not really usable for a gc-free `LogstashLayout`, hence I implemented `Layout<String>` directly myself. What I can further try is implementing `StringLayout` and see if that helps. I first need to reproduce the `KafkaAppender` problem locally. Need to get a Kafka Docker image, etc.

I will keep you updated.

**Comment by vy** [comment]
Sorry, totally my mistake. It should have been fixed with 72db860 that I've just pushed. Would you mind testing further, please? I really would like to ship a release for this branch.

**Comment by mkedwards** [comment]
Thanks; this looks like a clean, efficient way to provide the needed interface.  I'll apply our local changes (mostly just the `Streamable` interface) on top of it and verify.

**Comment by vy** [comment]
For the records, (as of date) `KafkaAppender` is not garbage-free, hence it requires a handle to the `byte[]` rather than implementing `ByteBufferDestination`.

**Comment by mkedwards** [comment]
Maybe we can fix that in KafkaAppender?

**Comment by vy** [comment]
Making a garbage-free `KafkaAppender` will probably necessitate a garbage-free Kafka client too. Sorry, I will pass that.

**Comment by mkedwards** [comment]
For what it's worth, I don't think it would be terribly difficult to extend the Kafka producer code to implement ByteBufferDestination.  There are half a dozen layers of plumbing on the way down to https://github.com/apache/kafka/blob/1.1.1/clients/src/main/java/org/apache/kafka/common/record/DefaultRecord.java#L175, but they all basically treat `byte[] value` as opaque.  It would be pretty straightforward to wrap `LogstashLayout.encode(LogEvent source, ByteBufferDestination destination)` in a lambda that captures `this` and `source` and adapts `ByteBufferDestination` to `java.io.DataOutputStream`.  The result would be assignable to `java.util.function.Consumer<java.io.DataOutputStream>`.  Plumb that through KafkaAppender to KafkaProducer and down to DefaultRecord, and we're done.

**Comment by vy** [comment]
@mkedwards, did you have chance to test in your side?

**Comment by vy** [comment]
Released v0.15. Let's address further discussion in another ticket.

**Comment by mkedwards** [comment]
Sorry I haven't updated this one in a bit.  We're running happily in QA with the code state at https://github.com/nexiahome/log4j2-logstash-layout/tree/eb35c1c208cf534bf493cc25946c700262f6ed2d , which is based on https://github.com/vy/log4j2-logstash-layout/commit/bbda80a56f8c5d4c94b3c0f06aad74a7c9d7e767 .  I'll rebase on v0.15 as released in advance of our canary in production (planned for Monday).
